Merge and partially refactor aircraft activities

reaperrr opened the pull request:
- merges all Heli* activities into the non-heli counterparts
- removes cases of "doing own stuff directly instead of using normal activities" from several places, mostly `Aircraft`
- ~removes `ReturnOnIdle` and `FlyAwayOnIdle` traits as well as `LandWhenIdle` aircraft property in favor of `ActionsWhenIdle`~ split to #15873.
- properly unties `CanHover` and `VTOL` so they work properly independently from each other

Notes for reviewers:
- there probably are more improvements that could be made to the now-merged activities, but I felt those should be left for follow-ups.
- I recommend reviewing (but not testing) commit by commit, but keep in mind that a few temporary work-arounds/hacks in earlier commits are replaced with proper solutions in later commits.

reaperrr committed:
Merged HeliLand into Land activity

Another small step towards unifying aircraft activities.

reaperrr committed:
Merge HeliFlyCircle into FlyCircle

Another step towards unifying aircraft activities

reaperrr committed:
Merge HeliReturnToBase into ReturnToBase

There was a lot of overlapping already, and no differences
a few 'if's couldn't handle.

reaperrr committed:
Merge HeliFly into Fly

and replace their *Idle derivatives with INotifyBecomingIdle implmentation on Aircraft.

reaperrr committed:
Merge HeliAttack into FlyAttack

reaperrr committed:
Remove Fly.AdjustAltitude

In favor of using updated Fly.FlyToward.

- Avoids some duplication.
- Makes VTOL take-off CanHover-agnostic.
- Improved FlyCircle:
  - turnSpeedOverride instead of hardcoded division
    (added IdleTurnSpeed to Aircraft to make use of it)
  - slimmed down code
  - this allowed to remove most AutoCarryall hacks

Additionally, The old copy-pasted code from HeliAttack
was still updating the actor facing and position directly,
instead of using FlyToward, which is changed with this commit.

reaperrr committed:
Add ActionsOnIdle to Aircraft

And remove all traits, properties and hard-codings
that would conflict with it.

A single list with possible actions is better than multiple,
possibly conflicting booleans and traits.
Designed to support fallbacks in case a listed action isn't
supported by that particular actor, to make life a little easier
for modders.

reaperrr commented:
Meh, I just realised this still lacks an update rule. Will add that later.

reaperrr commented:
Updated.

obrakmann commented on OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:
```diff
@@ -46,29 +64,76 @@ public override Activity Tick(Actor self)
+
+				Cancel(self);
+				self.SetTargetLine(newTarget, Color.Green);
+				return new Fly(self, newTarget);
```
```csharp
Queue(new Fly(self, newTarget));
return NextActivity;
```

obrakmann commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -66,12 +74,12 @@ public class AircraftInfo : ITraitInfo, IPositionableInfo, IFacingInfo, IMoveInf
 		public readonly bool CanHover = false;
 
+		[Desc("List of possible action to take when becoming idle in descending order of priority. Invalid entries (e.g. Hover when CanHover is false) will be skipped.")]
+		public readonly IdleAircraftAction[] ActionsWhenIdle = { IdleAircraftAction.Land, IdleAircraftAction.Hover, IdleAircraftAction.Circle };
```
I could see use cases where it would be nice to have an ActorInit for this, for example for aircraft spawned by scripts.

reaperrr commented on OpenRA.Mods.Common/Activities/Air/HeliAttack.cs:
```diff
@@ -76,7 +76,7 @@ public override Activity Tick(Actor self)
 			// If all valid weapons have depleted their ammo and Rearmable trait exists, return to RearmActor to reload and then resume the activity
 			if (rearmable != null && attackAircraft.Armaments.All(x => x.IsTraitPaused || !x.Weapon.IsValidAgainst(target, self.World, self)))
-				return ActivityUtils.SequenceActivities(new HeliReturnToBase(self, aircraft.Info.AbortOnResupply), this);
+				return ActivityUtils.SequenceActivities(new ReturnToBase(self, aircraft.Info.AbortOnResupply), this);
```
> Please do not re-use `this`, create a new instance instead.

Does this always apply, i.e. also where we currently `return this;`?

obrakmann commented:
No, `return this;` is fine.

reaperrr committed:
Fix queueing on aircraft activities

Adhere to Activity.cs guidelines.

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -606,10 +606,7 @@ public Activity MoveToTarget(Actor self, Target target)
-				return new Land(self, target);
-
-			return new HeliLand(self, false);
+			return !Info.VTOL ? new Land(self, target) : new Land(self);
```
Can we use the `self, target` version for both here?
If not, please drop the `!` and switch the cases to improve readability.

pchote commented on OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs:
```diff
@@ -1,135 +0,0 @@
-
-		public Actor ChooseResupplier(Actor self, bool unreservedOnly)
-		{
-			if (rearmable == null)
```
`ReturnToBase`'s `ChooseResupplier` does a TraitInfo query instead of checking against `rearmable` - GH won't let me comment there directly.

pchote commented on OpenRA.Mods.Common/Activities/Air/ReturnToBase.cs:
```diff
@@ -131,41 +136,98 @@ public override Activity Tick(Actor self)
+			else if (!isCalculated)
 				Calculate(self);
 
 			if (dest == null || dest.IsDead)
```
HeliRTB also checked `Reservable.IsReserved(dest)`. Is the omission here intentional?

reaperrr replied:
Actually no, this is no different from HeliRTB.

If you look at HeliRTB again, what you're referring to is basically what `TryToReserveResupplier` (and `Calculate` by calling it) above does now, and that one indeed checks whether the old dest is reserved, while only returning unreserved actors.

If `dest` isn't null or dead here, it means either `ChooseResupplier(self, true)` was successful in returning a valid, unreserved resupplier, or `dest` is carried over from a previous tick, in which case the only actor that can have reserved it should be this one.

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -746,15 +743,6 @@ public void ResolveOrder(Actor self, Order order)
 
 				UnReserve();
-
-				// TODO: Implement INotifyBecomingIdle instead
```
I know its a pain, but it would reduce the risk of this PR a lot if we could split this out into its own prereq PR.  This is introducing a plumbing change, which is quite different (wrt testing and types of regression) to the activity changes that make up the bulk of this PR.

reaperrr replied:
#15873

pchote commented on OpenRA.Mods.Common/Activities/Air/Fly.cs:
```diff
@@ -37,36 +37,37 @@ public Fly(Actor self, Target t, WDist minRange, WDist maxRange)
+				// If FlyToward is called from FlyCircle, the aircraft needs to use aircraft.Facing even if it CanHover,
+				// otherwise it will circle sideways.
+				if (aircraft.Info.CanHover && !isFlyCircle)
+					move = aircraft.FlyStep(flyBackward ? -desiredFacing : desiredFacing);
```
IMO it would be cleaner to define a `var direction = flyBackwards ? -1 : 1` and then use `FlyStep(direction * facing)` here and below.

pchote commented on OpenRA.Mods.Common/Activities/Air/FlyAttack.cs:
```diff
@@ -89,25 +89,35 @@ public override Activity Tick(Actor self)
+					return this;
+				}
+				else if (withinMaxRange && !withinMinRange && desiredFacing != aircraft.Facing)
+				{
```
Style nit: unnecessary braces

pchote commented on OpenRA.Mods.Common/Traits/Air/Aircraft.cs:
```diff
@@ -40,6 +40,9 @@ public class AircraftInfo : ITraitInfo, IPositionableInfo, IFacingInfo, IMoveInf
 		public readonly int TurnSpeed = 255;
 
+		[Desc("Turn speed to apply when aircraft flies in circles while idle. Defaults to TurnSpeed if negative.")]
+		public readonly int IdleTurnSpeed = -1;
```
Is there any feasible way to split this and e4691a0ca12bc2435dc5da03268194c664dbfd69 into their own prereq PR?

reaperrr replied:
#15848

reaperrr committed:
Adapt official mods to ActionsOnIdle

Ran update rules on them first and then cleaned up.

reaperrr committed:
Remove duplicate Aircraft from TS jumpjet husk

reaperrr committed:
Make ChooseResupplier not return dead actors

Lack of an IsDead check could  result in returning a dead actor,
even if alive alternatives were available.

reaperrr commented:
Updated and rebased onto #15873.

reaperrr commented:
By now, using it as reference to create smaller PRs from scratch is much easier than it would be to rebase this, and since it's linked from both #16131 and #16338 it's easy to find even if closed, so I think it's pointless to keep this open as it will be superseded by smaller PRs anyway.
